### DISCUSSION TITLE
getRequestSources() or how to trigger http requests with params

### DISCUSSION THREAD

**Discussion by colorcube** [state-adapt/state-adapt#37] [Q&A] [answered]
Here's an excerpt from the Books example:

```TypeScript
    const booksService = inject(BooksService);

    const booksRequestSources = getRequestSources(
      'books',
      booksService.fetchAll()
    );
    
    return adapt(['books', initialState, adapter], {
      receiveBooks: booksRequestSources.success$,
      receiveError: booksRequestSources.error$,
      ...
```      

I wrapped my stuff into a Service with this.store, but it's the same in principle.

My Http Service have a getWithQuery(query) method.
I want to trigger from outside getWithQuery() with my parameters.

No data should be fetched on startup (like Books example do).

Any hints?

**Comment by mfp22** [comment, marked as answer]
You can save the parameters as state. You can put them in the same store and use [watch](https://state-adapt.github.io/docs/rxjs#stateadaptwatch) to access that state without creating a circular reference. Or put them in another store and select as usual. Then filter to make sure parameters exist, switchMap to the request, then use [toRequestSource](https://state-adapt.github.io/docs/rxjs#torequestsource) and [spltRequestSources](https://state-adapt.github.io/docs/rxjs#splitrequestsources). Those docs have examples.

When an HTTP request doesn't need parameters, you can use [getRequestSources](https://state-adapt.github.io/docs/rxjs#getrequestsources). But when the HTTP request needs to be made in an inner observable, you need the 2 separate things mentioned above.

And obviously you don't actually need to store the parameters as state. That's just what I like to do. You could just have a subject that you call next on.

**Comment by colorcube** [comment]
Thanks. That helped. 
I wanted to store the parameters in the store anyway.

It was not obvious to me how the path looks like to be used with watch(). This works for me:

```Typescript
watch('mystore.query', queryAdapter).state$
```

**Comment by mfp22** [comment]
I was thinking more about this, and I think I may have given an outdated answer. I believe `getRequestSources`, which combines `toRequestSource` and `splitRequestSources`, is more flexible than it used to be. In the past, I had it named `getHttpSources` and `getHttpRequest` or something like that, and it needed to be involved in the HTTP observable itself. There was special logic for handling errors... and I realized it could be much more simple if I just required the HTTP observable itself to take care of throwing errors if needed. 

So yeah, I think you can usually just use `getRequestSources`, even if the HTTP request is an inner observable.

**Reply by xfh** [reply]
Hi @mfp22, can you please explain in which situations `getRequestSources` is a save choice when the HTTP request is an inner observable?

I've tried it, but after the first error, the stream is interrupted. Only when adding `toRequestSource` to the inner observable, it works as expected.

This works:
```js
    private createRequest = splitRequestSources('[Favoriten API] create', this.create$.pipe(
        concatMap(({payload}) => this.http.post(FavoritService.BASE_URL, payload).pipe(
            toRequestSource('[Favoriten API] create'))),
    ));
```

This works until the first error:
```js
    private createRequest = getRequestSources('[Favoriten API] create', this.create$.pipe(
        concatMap(({payload}) => this.http.post(FavoritService.BASE_URL, payload)),
    ));
```

In my example, `this.create$` is a source. The `createRequest.sucess$` and `createRequest.error$` observables are passed to state-adapt reactions.

If I produce an HTTP error and trigger the create$ action afterwards, the HTTP is no longer triggered. I wish there was a way to catch inner observable errors in the higher-level pipe and had hoped that you had found a way with some magic in `getRequestSources` :-)

**Reply by mfp22** [reply]
Oh... So maybe you do need to split them up like that then. I don't remember.

You can see in the source code for getRequestSources that it just wraps the other two function calls anyway. https://github.com/state-adapt/state-adapt/blob/main/libs/rxjs/src/lib/sources/get-request-sources.function.ts

**Reply by mfp22** [reply]
I didn't realize if an inner observable threw that it would interrupt the whole stream. My previous answer (to which these are replies) did not take that into account. In that comment I was referring to the HTTP observable checking for status 404 (for example) and throwing internally. I didn't realize it needed to be handled before the outer observable could receive it.

So yes, you need toRequestSource and splitRequestSources.

**Reply by xfh** [reply]
Thanks for clarifying!
